feat: emit download progress and allow changing installation path

esakal opened the pull request:
Hi @juliangruber thanks for creating this library. This PR was created to support packaged application like electron where the node_modules is bundled and compressed and it is not possible for a library to simply download to its' root.

# Goals
## customize installation path
I added a new option named `installPath` which default to `__dirname` to be backward compatible. I also updated the `readme.md` file with a suggestion about relevant value if using electron to pack this library.

## report download progress
As applications might want to show download progress I propagated the `onProgress` event from `got` library thru the options. This is optional and will work also if the user choose not to monitor the progress.

juliangruber commented on README.md:
```diff
@@ -50,10 +50,20 @@ $ npm install download-chromium
+	revision = '499413', 
+	log = false, 
+	onProgress = undefined, 
+	installPath = '{root folder od download-chromium module}' })
```
```suggestion
	installPath = rootFolderOfThisModule
})
```

esakal replied:
done, following other comment I modified it to `'{rootFolderOfThisModule}/.local-chromium'`

juliangruber commented on README.md:
````diff
@@ -50,10 +50,20 @@ $ npm install download-chromium
+```
 Returns a Promise resolving with the Chromium executable path.
 
+* use `installPath` if running from within `electron` packaged app (for example set install path to `require('electron').app.getPath('userData')`)
````
```suggestion
* `installPath`: set the install path if you can't write your `node_modules`, eg from within an `electron` packaged app (`installPath: require('electron').app.getPath('userData')`)
```

juliangruber commented on README.md:
```diff
@@ -50,10 +50,20 @@ $ npm install download-chromium
 Returns a Promise resolving with the Chromium executable path.
 
+* use `installPath` if running from within `electron` packaged app (for example set install path to `require('electron').app.getPath('userData')`)
+* provide `onProgress` to track download progress. `onProgress` gets single argument: `{ percent, transferred, total }` 
```
```suggestion
* `onProgress`: track download progress. `onProgress` receives one argument `{ percent, transferred, total }` 
```

juliangruber commented on index.js:
```diff
@@ -98,10 +101,13 @@ const copyCacheToModule = async (moduleExecutablePath, platform, revision) => {
+                          log: log = false,
+                          installPath: installPathBase = __dirname,
+                          onProgress: onProgress
+                        } = {}) => {
```
indentation is off, please stick to 2 spaces

esakal replied:
done

juliangruber commented on index.js:
```diff
@@ -98,10 +101,13 @@ const copyCacheToModule = async (moduleExecutablePath, platform, revision) => {
+                          installPath: installPathBase = __dirname,
+                          onProgress: onProgress
+                        } = {}) => {
+  const installPath = `${installPathBase}/.local-chromium`
```
this isn't what I'd expect, if I set `installPath` that's where it should be downloaded to, not `${installPath}/.local-chromium`. I'd suggest rather to default `installPath` to `${__dirname}/.local-chromium`.

esakal replied:
I agree, done

esakal commented:
Hi @juliangruber sorry for the far to late reply :/ 
- [x] I implemented all you comments 
- [x] I adjusted test script to use onProgress as well

> 1. `onProgress` is perfect, I'd merge that right away
Great 👍 

> Also, CI is failing. Can you take a look?

can you add a link to travis log? because running the following works on my machine:
```
npm ci
npm test
```

> 2. I'm not sure if making `installPath` configurable is necessary, what about instead of writing to `${__dirname}/.local-chromium`, we could write to `~/.local-chromium` or something.
> 

1. using `~/.local-chromium` or any predefined path without allowing to manually modify it will fail if two different applications are using different `puppeteer-core` versions thus having different local chromium versions. The library must allow people to isolate their version if needed.
2. `{__dirname}` is not relevant for Electron applications that are using ASAR file (which is enabled by default) as `download-chromium` library is bundled into the Electron package in a read only area.

Let me know what you think. This time I promise to be responsive :)

juliangruber commented on README.md:
````diff
@@ -50,10 +50,20 @@ $ npm install download-chromium
+```
 Returns a Promise resolving with the Chromium executable path.
 
+* `installPath`: set the install path if you can't write your `node_modules`, eg from within an `electron` packaged app (`installPath: require('electron').app.getPath('userData')`)
````
```suggestion
* `installPath`: set the install path if you can't write to your `node_modules`, eg from within an `electron` packaged app (`installPath: require('electron').app.getPath('userData')`)
```

juliangruber commented on README.md:
```diff
@@ -50,10 +50,20 @@ $ npm install download-chromium
+	revision = '499413', 
+	log = false, 
+	onProgress = undefined, 
+	installPath = '{rootFolderOfThisModule}/.local-chromium' })
```
```suggestion
	installPath = '{__dirname}/.local-chromium' })
```

juliangruber commented on bin.js:
```diff
@@ -3,7 +3,17 @@
 const download = require('.')
 
-download({ log: true })
+function onPregress ({ percent, transferred, total }) {
```
```suggestion
function onProgress ({ percent, transferred, total }) {
```

juliangruber commented on bin.js:
```diff
@@ -3,7 +3,17 @@
+}
+download({
+  log: true,
+  onProgress: onPregress
```
```suggestion
  onProgress
```

esakal commented:
> almost!!

Great ;) 

Regarding travis, I checked it again. It seems to fail when building on windows. This is something that happens in the library prior to my change. see [this](https://travis-ci.org/juliangruber/download-chromium/builds/532270179) for example.

juliangruber commented:
> Regarding travis, I checked it again. It seems to fail when building on windows. This is something that happens in the library prior to my change. see [this](https://travis-ci.org/juliangruber/download-chromium/builds/532270179) for example.

You're right :|

Thank you for this! Published as 2.2.0 :)
